Documentation/Development.md

```diff
@@ -1,37 +1,103 @@
-# Swift for TensorFlow Deep Learning Library
+# Swift for TensorFlow Developer Guide
 
-## Development
+## Requirements
 
-### Requirements
+* The latest Swift toolchain snapshot from [swift.org][swift]
 
-* The latest Swift toolchain snapshot from [swift.org](https://swift.org/download/#sanpshots).
+Swift for TensorFlow APIs can be built with either:
 
-### Building
+* CMake 3.16 or later from [cmake.org][cmake]
+* Swift Package Manager (included in the above toolchain)
 
-### X10
+## Components
 
-X10 provides the underlying lazy tensor implementation that is used for the
-tensor computation in the TensorFlow Swift APIs. This library builds atop of
-the TensorFlow library, using the XLA compiler to perform optimizations over
-the tensor computation.
+Swift for TensorFlow APIs are composed of two distinct components:
 
-#### (Option 1) Use a prebuilt version of X10
+1. X10
+2. Swift APIs
 
-You can use a prebuilt version of the X10 library for building the TensorFlow
-Swift APIs package if you do not need to make changes to the X10 library
+<details>
+    <summary>What is X10?</summary>
+
+> X10 provides the underlying lazy tensor implementation that is used for
+> tensor computations in the Swift for TensorFlow APIs. This library builds on
+> top of the TensorFlow library, using the XLA compiler to perform
+> optimizations.
+
+</details>
+
+
+The two components can be built together (CMake only) or separately.
+
+Follow the instructions below based on your preferred build tool.
+
+## Building With CMake
+
+> **Note:** In-tree builds are not supported.
+
+> **Note:** To enable CUDA support, run `export TF_NEED_CUDA=1` before the
+  steps below.
+
+> **Note:** If `swiftc` is not in your `PATH`, you must specify the path to it
+  using `-D CMAKE_Swift_COMPILER=`.
+
+This will build both X10 and Swift APIs.  Ensure that you do not have the
+X10 modules in the toolchain that you are using to develop here.
+
+```shell
+cmake -B out -G Ninja -S swift-apis
+cmake --build out
+```
+
+To run tests:
+
+> **Note:** To view failure output, run `export CTEST_OUTPUT_ON_FAILURE=1`
+  before running tests.
+
+```shell
+cmake --build out --target test
+```
+
+### macOS
+
+On macOS, passing `-D BUILD_TESTING=NO` is currently necessary to skip building
+tests. This avoids an error: `cannot load underlying module for 'XCTest'`.
+
+```shell
+cmake -B out -D USE_BUNDLED_CTENSORFLOW=YES -D USE_BUNDLED_X10=YES -D BUILD_TESTING=NO -G Ninja -S swift-apis
+cmake --build out
+```
+
+## Building With Swift Package Manager
+
+Building with SwiftPM involves building X10 and Swift APIs separately.
+
+### Building X10
+
+To determine the appropriate build path, first consider whether you are
+modifying X10.
+
+* If you are modifying files in any subdirectory within **Sources/CX10** or
+  **Sources/x10**, you are modifying X10 and must follow **Option 2**.
+* Otherwise, use **Option 1**.
```

Comment:
I think that we need to be more conservative about Option 1.  We should explicitly call out, if you are not building for Windows x64 or macOS x86_64, then you must use Option 2.  This is the reason that I had the options ordered as such, Option 1 is a special case, and you quickly fall back to option 2.  I don't have a strong opinion on how we name it order it, but I do feel like we need to be overly explicit that Option 1 is a very narrow set of builds.

Reply:
I'd like to surface the least complex path first. I added more detail, but it still feels a bit convoluted. I'll keep thinking of a better way to organize this.